components/num-util/src/fixed_decimal.rs

```diff
@@ -0,0 +1,517 @@
+    /// magnitude of the most significant digit is std::i16::MAX, and the lowest magnitude of the
+    /// least significant digit is std::i16::MIN.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use icu_num_util::FixedDecimal;
+    /// use icu_num_util::fixed_decimal::Error;
+    ///
+    /// let mut dec1 = FixedDecimal::from(123);
+    /// assert_eq!(Error::Limit, dec1.adjust_magnitude(std::i16::MAX).unwrap_err());
+    /// ```
+    Limit,
+}
+
+/// A struct containing decimal digits with efficient iteration and manipulation by magnitude
+/// (power of 10). Supports a mantissa of non-zero digits and a number of leading and trailing
+/// zeros, used for formatting and plural selection.
+///
+/// You can create a FixedDecimal from an unsigned integer or an isize.
+///
+/// # Examples
+///
+/// ## From an isize
+///
+/// ```
+/// use icu_num_util::FixedDecimal;
+///
+/// let dec1 = FixedDecimal::from(250);
+/// assert_eq!("250", dec1.to_string());
+/// ```
+///
+/// ## From a BigUint
+///
+/// ```
+/// use icu_num_util::FixedDecimal;
+/// use icu_num_util::fixed_decimal::UnsignedWrapper;
+/// use num_bigint::BigUint;
+/// use std::convert::TryFrom;
+///
+/// let input = BigUint::parse_bytes(b"98765432109876543210", 10).unwrap();
+/// let dec2 = FixedDecimal::try_from(UnsignedWrapper(input)).unwrap();
+/// assert_eq!("98765432109876543210", dec2.to_string());
+/// ```
+#[derive(Debug, Clone, PartialEq)]
+pub struct FixedDecimal {
+    /// List of digits; digits[0] is the most significant.
+    ///
+    /// Invariants:
+    /// - Must not include leading or trailing zeros
+    /// - Length must not exceed (magnitude - lower_magnitude + 1)
+    // TODO: Consider using a nibble array
+    digits: SmallVec<[u8; 8]>,
+
+    /// Power of 10 of digits[0].
+    ///
+    /// Invariants:
+    /// - <= upper_magnitude
+    /// - >= lower_magnitude
+    magnitude: i16,
+
+    /// Power of 10 of the most significant digit, which may be zero.
+    ///
+    /// Invariants:
+    /// - >= 0
+    /// - >= magnitude
+    upper_magnitude: i16,
+
+    /// Power of 10 of the least significant digit, which may be zero.
+    ///
+    /// Invariants:
+    /// - <= 0
+    /// - <= magnitude
+    lower_magnitude: i16,
+
+    /// Whether the number is negative. Negative zero is supported.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use icu_num_util::FixedDecimal;
+    ///
+    /// let mut dec: FixedDecimal = Default::default();
+    /// dec.is_negative = true;
+    /// assert_eq!("-0", dec.to_string());
+    /// ```
+    pub is_negative: bool,
+}
+
+impl Default for FixedDecimal {
+    /// Returns a FixedDecimal representing zero.
+    fn default() -> Self {
+        FixedDecimal {
+            digits: SmallVec::new(),
+            magnitude: 0,
+            upper_magnitude: 0,
+            lower_magnitude: 0,
+            is_negative: false,
+        }
+    }
+}
```

Comment:
How common do you expect someone to want to create a `FixedDecimal` of value `0`?

I imagine that mostly being useful if there's a struct that has a field of `FixedDecimal` type and we'd want to have `Default` derived for it.
But that feels weird - after all, if someone wants a `FixedDecimal` they're most likely gonna accept some input and parse it into `FixedDecimal`, no? What would a default value of `0` give, if they still need/want to later take that input and reassign the field?

Reply:
Not sure what you're suggesting here.  Are you suggesting to not implement the `Default` trait because you question whether it will be widely used?

I implemented the trait because it seems useful for a data type to have a Default trait as its only constructor.  Also, the `from` methods call `Default` before appending things to the digit vector.  Without this trait, I would probably still have a method returning the same thing, but I could make it private.

Reply:
I'm trying to understand the what `FixedDecimal` is, and it seems to me a particular type of `Number` with special methods.
You wouldn't have `Number:;default()` method, right? What would it mean? If it existed, then it'd likely return `0` for `Default`, but it most likely would not exist.

That's why I'm questioning whether we should have `Default` trait on it at all.

As to what you're calling in `from`, you can just construct it manually.

Reply:
I think 0 is a very sensible default for FixedDecimal: it's a decimal number with no significant digits.

Reply:
> I'm trying to understand the what FixedDecimal is

Here are some attempts at answering this question.  Hopefully one of them sticks:

1. A FixedDecimal is a base-10 number in which every position (power of 10) is mapped to a decimal digit between 0 and 9.
2. A FixedDecimal is a base-10 number with zero or more significant (nonzero) digits.
3. A FixedDecimal is a numbering-system-agnostic representation of a base 10 number as humans would think about it.